airbyte-workers/src/main/java/io/airbyte/workers/DefaultReplicationWorker.java

```diff
@@ -95,21 +106,59 @@ public StandardSyncOutput run(StandardSyncInput syncInput, Path jobRoot) throws
         destination.start(destinationConfig, jobRoot);
         source.start(sourceConfig, jobRoot);
 
-        while (!cancelled.get() && !source.isFinished()) {
-          final Optional<AirbyteMessage> maybeMessage = source.attemptRead();
-          if (maybeMessage.isPresent()) {
-            final AirbyteMessage message = mapper.mapMessage(maybeMessage.get());
-
-            messageTracker.accept(message);
-            destination.accept(message);
+        final Thread sourceThread = new Thread(() -> {
+          try {
+            while (!cancelled.get() && !source.isFinished()) {
+              final Optional<AirbyteMessage> messageOptional = source.attemptRead();
+              if (messageOptional.isPresent()) {
+                final AirbyteMessage message = mapper.mapMessage(messageOptional.get());
+
+                sourceMessageTracker.accept(message);
+                destination.accept(message);
```

Comment:
agreed. For now the queue doesn't really add anything since the backpressure is applied by `destination.accept` blocking. I was going to hold off on doing this until we actually needed the queue, but I can definitely just do it now.

Reply:
Agreed. The blocking queue wouldn't do anything for now, since `destination.accept` applies the back pressure for us. I was going to wait to add the queue until we actually needed it. I can just go ahead and do it now.

Reply:
I vote for leaving it out until we actually need it.